### PR TITLE
Make iseq eval consistent with Kernel eval

### PR DESCRIPTION
**Pull request by dalehamel**
This should fix https://bugs.ruby-lang.org/issues/12093, and is an additive change only.

This allows this code to work:

```ruby
obj = Struct.new(:a, :b).new(1, 2)
bind = obj.instance_eval {binding}
RubyVM::InstructionSequence.compile("a + b").eval(bind) #=> 3
```

As I've verified with `make test-all TESTS=test/ruby/test_iseq.rb`

I think this should be added to help make the API with Kernel.eval be more consistent:
* Using Kernel.eval, you can specify source code and a binding to evaluate it within
* Using RubyVM::InstructionSequence, you can compile source code to an instruction sequence
* You can execute source code within a binding, but you cannot execute compiled instruction sequences against a binding, so the features of Kernel.eval differ from iseq.eval

To fix this, the signature for iseq.eval is changed to allow for 0-1 arguments to be passed. I see this as keeping the same signature as `Kernel.eval`, but just dropping the first parameter as the target to evaluate is implicit as `self`, which I believe helps to keep the API consistent.

This new functionality should not add any performance penalty or code changes because 0 arguments are passed, the current behavior remains, and the current signature is preserved.  If one argument is passed, then it is used as the binding, and attempts to mimic the behavior of [vm_eval](https://github.com/ruby/ruby/blob/master/vm_eval.c#L1350-L1368 ) in its handling of the binding.

While the originally introduced `eval_with` can achieve the same thing, I think that doing this with `eval` is more consistent with the rest of the ruby API, and with Kernel.eval. The difference being that the thing to evaluate (the instruction sequence) is explicit via `self` rather than the first required parameter. This is the value I think over the original patch.

Based on original patch: https://github.com/ruby/ruby/compare/trunk...nobu:feature/12093-iseq-eval_with, but I've updated it follow the conventions of https://github.com/ruby/ruby/blob/master/vm_eval.c#L1350-L1368 

I've added a basic test that accepts the binding to the eval call, and verified the existing test continue to pass as i've only extended, not changed, the API here.

I have also updated the doc header for the method, and attempted to keep it consistent with the existing docs for Kernel.eval.

@nobu can you please take a look as this is based on your original patch and discussion in redmine.

Thanks @methodmissing for preliminary review

### CODE REVIEW

File: iseq.c [hunk 1]
```diff
@@ -1332,17 +1332,33 @@ rb_iseqw_to_iseq(VALUE iseqw)
...
  */
 static VALUE
-iseqw_eval(VALUE self)
+iseqw_eval(int argc, const VALUE *argv, VALUE self)
 {
-    rb_secure(1);
-    return rb_iseq_eval(iseqw_check(self));
+    VALUE scope;
+
+    if (argc == 0) {
```
Comment (nobu) [inline iseq.c]: `Kernel#eval` accepts as `nil` as the binding, so this method too, I think.

File: iseq.c [hunk 2]
```diff
@@ -1332,17 +1332,33 @@ rb_iseqw_to_iseq(VALUE iseqw)
...
+        rb_secure(1);
+        return rb_iseq_eval(iseqw_check(self));
+    }
+    else {
+        rb_scan_args(argc, argv, "01", &scope);
```
Comment (nobu) [inline iseq.c]: With `rb_check_arity(argc, 0, 1)` then `rb_secure(1)`, duplicate code can be reduced.

File: vm.c [hunk 3]
```diff
@@ -2137,6 +2137,22 @@ rb_iseq_eval(const rb_iseq_t *iseq)
...
+rb_iseq_eval_in_scope(const rb_iseq_t *iseq, VALUE scope)
+{
+    rb_execution_context_t *ec = GET_EC();
+    rb_binding_t *bind = Check_TypedStruct(scope, &ruby_binding_data_type);
+
+    vm_set_eval_stack(ec, iseq, NULL, &bind->block);
+
+    /* save new env */
+    if (iseq->body->local_table_size > 0) {
+      vm_bind_update_env(scope, bind, vm_make_env_object(ec, ec->cfp));
```
Comment (nobu) [inline vm.c]: The indent is slipped.

**Comment by nobu** [comment]
Please add the reference to the ticket in the commit log.

File: iseq.c (hunk 2 above)
Comment (dalehamel) [reply iseq.c]: Yes I agree, I'll update it to scan check for the arity and refactor it accordingly

File: iseq.c (hunk 1 above)
Comment (dalehamel) [reply iseq.c]: Yes I agree, I checked the `Kernel#eval` implementation, and i think that just evaluating as if no binding was passed is the best behavior, so I put it in the same if branch

File: vm.c (hunk 3 above)
Comment (dalehamel) [reply vm.c]: My apologies for the style error, I've fixed this.

**Comment by dalehamel** [comment]
> Please add the reference to the ticket in the commit log.

I've added the bug URL to the commit message.

I believe I've addressed all comments, I've asked some of my colleagues to take a look as well.

**Comment by dalehamel** [comment]
I realized that I missed a flaw described in the original issue https://bugs.ruby-lang.org/issues/12093#note-11 so this feature has problems when a local variable and method call have the same name, with the method call overshadowing a local variable with the same identifier from the binding, giving unexpected results.

**Comment by dalehamel** [comment]
The ticket is rejected

**Pull request closed without merging**